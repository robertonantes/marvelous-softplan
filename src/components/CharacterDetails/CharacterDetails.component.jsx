import React from "react";
import {
  Wrapper,
  Details,
  Name,
  Description,
  ImageContainer,
  Image,
  AppearsOn,
} from "./CharacterDetails.styles.js";

const CharacterDetails = ({ data }) => {
  const { name, description, comics, series, stories, events } = data;
  return (
    <Wrapper data-testid="character-details">
      <Details>
        <Name>{name}</Name>
        <Description>{description || "No description available"}</Description>
        <AppearsOn>
          <p data-testid="comics">
            Comics: <span>{comics.available}</span>
          </p>
          <p data-testid="events">
            Events: <span>{events.available}</span>
          </p>
          <p data-testid="series">
            Series: <span>{series.available}</span>
          </p>
          <p data-testid="stories">
            Stories: <span>{stories.available}</span>
          </p>
        </AppearsOn>
      </Details>
      <ImageContainer>
        <Image
          src={`${data.thumbnail.path}.${data.thumbnail.extension}`}
          width="420px"
          height="420px"
          data-testid="character-image"
        />
      </ImageContainer>
    </Wrapper>
  );
};
export default CharacterDetails;
