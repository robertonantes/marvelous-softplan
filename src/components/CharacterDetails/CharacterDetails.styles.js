import styled from "styled-components";

export const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const Details = styled.div`
  position: relative;
`;

export const Name = styled.h1`
  color: ${({ isEditing }) => (isEditing ? "transparent" : "#fff")};
  font-size: 48px;
  text-decoration: underline;
  margin-top: 0;
  margin-bottom: 16px;
  ::selection {
    color: transparent;
  }
`;

export const Description = styled.p`
  color: #b9b9b9;
  font-weight: 300;
  font-size: 16px;
  line-height: 21px;
  max-width: 530px;
`;

export const AppearsOn = styled.div`
  margin-top: 40px;

  p {
    color: #fff;
    font-weight: 500;
    margin-bottom: 8px;
    margin-top: 0;

    span {
      font-weight: 300;
    }
  }
`;

export const ImageContainer = styled.div``;
export const Image = styled.img`
  max-width: 420px;
  border-radius: 50%;
`;

export const EditBtn = styled.button`
  color: #000;
  background-color: #fff;
  border: 0;
  border-radius: 4px;
  font-weight: 300;
  cursor: pointer;
  outline: 0;
`;
