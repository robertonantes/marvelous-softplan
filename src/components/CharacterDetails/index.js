export { default } from "./CharacterDetails.component";
