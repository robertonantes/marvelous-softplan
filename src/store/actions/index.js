import {
  CHARACTERS_LOADED,
  FETCH_CHARACTERS,
  SEARCH_CHARACTERS,
  SEARCH_COMPLETED,
} from "./types";
import { fetchCharacters } from "../../common/services";

export function fetchLatestCharacters() {
  return async function (dispatch) {
    dispatch({ type: FETCH_CHARACTERS });
    const options = {
      orderBy: "-modified",
      limit: 20,
    };
    const response = await fetchCharacters(options);
    dispatch({ type: CHARACTERS_LOADED, payload: response.data });
  };
}

export function searchCharacters(term) {
  return async function (dispatch) {
    dispatch({ type: SEARCH_CHARACTERS });

    const options = {
      nameStartsWith: term,
      orderBy: "name",
      limit: 5,
    };

    const response = await fetchCharacters(options);
    dispatch({ type: SEARCH_COMPLETED, payload: response.data });
  };
}
