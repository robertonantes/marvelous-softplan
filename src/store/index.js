import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";

import rootReducer from "./reducers";
import { cacheLatestCharacters } from "./middlewares/cache.middleware";

export default createStore(
  rootReducer,
  applyMiddleware(thunk, cacheLatestCharacters)
);
