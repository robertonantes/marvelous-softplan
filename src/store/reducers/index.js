import { combineReducers } from "redux";
import characters from "./characters.reducer";
import search from "./search.reducer";

export default combineReducers({ characters, search });
