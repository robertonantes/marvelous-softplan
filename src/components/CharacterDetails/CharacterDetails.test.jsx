import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import CharacterDetails from ".";
import store from "../../store";

function mockData(data = {}) {
  return {
    name: "Character Name",
    description: "Character description",
    comics: {
      available: 10,
    },
    series: {
      available: 11,
    },
    stories: {
      available: 12,
    },
    events: {
      available: 13,
    },
    thumbnail: {
      path: "imagepath",
      extension: "jpg",
    },
    ...data,
  };
}

describe("CharacterDetails Component", () => {
  it("Renders description prop", () => {
    render(
      <Provider store={store}>
        <CharacterDetails data={mockData()} />
      </Provider>
    );
    expect(screen.getByText(/character description/i)).toBeInTheDocument();
  });

  it("Renders Â´no description available' when description prop is falsy", () => {
    render(
      <Provider store={store}>
        <CharacterDetails data={mockData({ description: "" })} />
      </Provider>
    );
    expect(screen.getByText(/no description available/i)).toBeInTheDocument();
  });

  it("Renders character info", () => {
    const data = mockData();
    const { asFragment } = render(
      <Provider store={store}>
        <CharacterDetails data={data} />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
