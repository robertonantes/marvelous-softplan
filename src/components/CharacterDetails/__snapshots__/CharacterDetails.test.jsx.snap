// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetails Component Renders character info 1`] = `
<DocumentFragment>
  <div
    class="sc-AxiKw cGtQMs"
    data-testid="character-details"
  >
    <div
      class="sc-AxhCb jZxptO"
    >
      <h1
        class="sc-AxhUy hTMfeP"
      >
        Character Name
      </h1>
      <p
        class="sc-AxgMl iaKBpG"
      >
        Character description
      </p>
      <div
        class="sc-AxheI dXqNRb"
      >
        <p
          data-testid="comics"
        >
          Comics: 
          <span>
            10
          </span>
        </p>
        <p
          data-testid="events"
        >
          Events: 
          <span>
            13
          </span>
        </p>
        <p
          data-testid="series"
        >
          Series: 
          <span>
            11
          </span>
        </p>
        <p
          data-testid="stories"
        >
          Stories: 
          <span>
            12
          </span>
        </p>
      </div>
    </div>
    <div
      class="sc-Axmtr hBVTto"
    >
      <img
        class="sc-AxmLO hypiLP"
        data-testid="character-image"
        height="420px"
        src="imagepath.jpg"
        width="420px"
      />
    </div>
  </div>
</DocumentFragment>
`;
